### ISSUE TITLE
[validation] changing the preferences doesn't have any immediate effect

### ISSUE THREAD

**Issue by martinlippert** [spring-projects/sts4#833]
When I enable the open-rewrite base reconciling and have a class with an `@Autowired` annotated constructor, this is flagged as a warning (which seems to be correct). When I go to the preferences and change them (e.g. to switch off the boot 2.x validations), nothing happens. The warning stays around (within the open editor as well as inside of the problems view).

This behavior doesn't seem to be specific for this specific validation, but a general issue.

**Comment by martinlippert** [comment]
The changes in the preferences seem to take effect when the language server is restarted (e.g. after closing the last open editor, waiting for the timeout, and re-opening an editor after the timeout).

**Comment by BoykoAlex** [comment]
Hmm... don't think I'm seeing this at all :-\ Will try VSCode...

**Comment by martinlippert** [comment]
I was trying the latest CI build of STS 4.16.0 on Eclipse 2022-09...

**Comment by martinlippert** [comment]
Hmmm... I tried this again, here is what I found:

First, I tried the latest CI build on Eclipse 2022-09, created a RestService guide example, played with it, added classes and added a constructor with that `@Autowired` annotation. Then I switched on the open-rewrite reconciling preference, the warning on the annotation appeared and I was able to switch the validation on and off as well as changed the severity of it several times while having the file open in an editor and while not having the file open in an editor (opened another one from the same project). All that worked just fine, the validation result changed just fine.

Then, I changed the class a bit to try the other quick fix that we have to turn autowired fields into constructors, all that worked as well. Then, I added the `@Autowired` annotation to the generated constructor again and the warning appeared on it. But changing the validation preferences didn't have any effect anymore, I was not able to get rid of the warning at all.

I shutdown the language server, waited for the timeout, and enabled the log output for it, and here is what I can see right after the language server started as well as when I try to change the preferences. No idea how it got into this somehow broken state, but that seems to prevent the stuff from working somehow.

```
12:23:24.756 [pool-10-thread-1] WARN  o.e.lsp4j.jsonrpc.RemoteEndpoint - Notification threw an exception: {
  "jsonrpc": "2.0",
  "method": "workspace/didChangeConfiguration",
  "params": {
    "settings": {
      "spring-boot": {
        "ls": {
          "problem": {
            "boot2": {
              "JAVA_AUTOWIRED_CONSTRUCTOR": "INFO"
            }
          }
        }
      },
      "boot-java": {
        "support-spring-xml-config": {
          "scan-folders": "src/main",
          "content-assist": "true",
          "hyperlinks": "true",
          "on": false
        },
        "remote-apps": [],
        "scan-java-test-sources": {
          "on": false
        },
        "live-information": {
          "automatic-tracking": {
            "delay": 5000,
            "on": true
          },
          "fetch-data": {
            "retry-delay-in-seconds": 3,
            "max-retries": 10
          }
        },
        "change-detection": {
          "on": false
        },
        "validation": {
          "spel": {
            "on": false
          }
        },
        "rewrite": {
          "reconcile": true
        }
      }
    }
  }
}
java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:67)
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.notify(GenericEndpoint.java:152)
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.handleNotification(RemoteEndpoint.java:220)
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.consume(RemoteEndpoint.java:187)
	at org.springframework.ide.vscode.commons.languageserver.util.ParentProcessWatcher.lambda$apply$0(ParentProcessWatcher.java:129)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.handleMessage(StreamMessageProducer.java:194)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.listen(StreamMessageProducer.java:94)
	at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor.run(ConcurrentMessageProcessor.java:113)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.lang.reflect.InvocationTargetException: null
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:65)
	... 12 common frames omitted
Caused by: java.lang.IllegalArgumentException: No enum constant org.springframework.ide.vscode.commons.languageserver.reconcile.ProblemCategory.Toggle.Option.false
	at java.base/java.lang.Enum.valueOf(Enum.java:273)
	at org.springframework.ide.vscode.commons.languageserver.reconcile.ProblemCategory$Toggle$Option.valueOf(ProblemCategory.java:59)
	at org.springframework.ide.vscode.boot.app.BootJavaConfig.isSpelExpressionValidationEnabled(BootJavaConfig.java:129)
	at org.springframework.ide.vscode.boot.java.BootJavaLanguageServerComponents.lambda$new$0(BootJavaLanguageServerComponents.java:226)
	at org.springframework.ide.vscode.commons.languageserver.util.ListenerList.fire(ListenerList.java:23)
	at org.springframework.ide.vscode.boot.app.BootJavaConfig.handleConfigurationChange(BootJavaConfig.java:151)
	at org.springframework.ide.vscode.commons.languageserver.util.ListenerList.fire(ListenerList.java:23)
	at org.springframework.ide.vscode.commons.languageserver.util.SimpleWorkspaceService.didChangeConfiguration(SimpleWorkspaceService.java:81)
	... 17 common frames omitted
```

@BoykoAlex Any idea what is going on here?

And, most likely, this is the same issue as https://github.com/spring-projects/sts4/issues/834 indeed.

**Comment by martinlippert** [comment]
But I can't easily reproduce this problem. Trying again and it doesn't happen. I very much hope that you can make some sense out of the exception @BoykoAlex

**Comment by BoykoAlex** [comment]
This has to do with legacy setting for SPEL expression validation on/off which used to be a boolean string `true` or `false` (or maybe upper case) and became an enum string `ON` or 'OFF`. I had the backward compatibility code for `true` but not for `false`. Note that after switching this setting once it'll become the new enum string value and the issue will stop occurring. 
The other issue is that listeners are notified about the config change. One of them fails with the exception above and effectively short circuits the loop over listeners. Consequently listeners triggering reconciling are not reached resulting in same problems still showing with the same severity. I fixed this by wrapping listeners loop body in a try-catch block.

**Comment by BoykoAlex** [comment]
Fixed with b4ed5bc68d2f889574bf84bc76c4c2a72066622e
